### PR TITLE
Added decoding support for FSCTL_QUERY_NETWORK_INTERFACE_INFO

### PR DESCRIPTION
**Pull request by sudosantanu**
Added decoding support for FSCTL_QUERY_NETWORK_INTERFACE_INFO

### CODE REVIEW

**Comment by isi-adas** [comment]
cc: @isi-rszczesniak , @marchhare09

File: src/pike/smb2.py [hunk 1]
```diff
@@ -4169,8 +4181,40 @@ def __init__(self, parent):
         IoctlOutput.__init__(self, parent)
 
     def _decode(self, cur):
-        pass
-
+        self.next_loc ={}
```
Comment (isi-adas) [inline src/pike/smb2.py:4184]: Spacing needs consistency on both sides of `=`.

File: src/pike/smb2.py [hunk 2]
```diff
@@ -4169,8 +4181,40 @@ def __init__(self, parent):
...
+        self.port = {}
+        self.ip_address = {}
+        self.flow_info = {}
+        self.scope_id = {}
+        self.reserved_2 ={}
+        self.reserved_3 = {}
+
+        i = 0
+        while True :
+                self.next_loc[i] = cur.decode_uint32le()
```
Comment (isi-adas) [inline src/pike/smb2.py:4199]: Indentations would be good to be 4 spaces.

File: src/pike/smb2.py [hunk 3]
```diff
@@ -4169,8 +4181,40 @@ def __init__(self, parent):
...
+                        self.reserved_2[i] = cur.decode_uint64le()
+                        self.reserved_3[i] = cur.decode_bytes(112)
+                if self.family[i] == IPv6 : #IPv6
+                        self.flow_info[i] = cur.decode_uint32le()
+                        self.ip_address[i] =  binascii.hexlify(cur.decode_bytes(16))
+                        self.scope_id[i] = cur.decode_uint32le()
+                        self.reserved_3[i] = cur.decode_bytes(100)
+                if self.next_loc[i] == 0:
+                        break
+                i = i+1
```
Comment (isi-adas) [inline src/pike/smb2.py:4217]: Spaces around `+`.

File: src/pike/smb2.py [hunk 4]
```diff
@@ -4169,8 +4181,40 @@ def __init__(self, parent):
...
+        self.capability = {}
+        self.reserved = {}
+        self.link_speed = {}
+        self.family = {}
+        self.port = {}
+        self.ip_address = {}
+        self.flow_info = {}
+        self.scope_id = {}
+        self.reserved_2 ={}
+        self.reserved_3 = {}
```
Comment (isi-adas) [inline src/pike/smb2.py:4195]: Would it be nice or correct to move this entire chunk to __init__() ?

**Review by isi-adas — APPROVED** [review]

File: src/pike/smb2.py [hunk 5]
```diff
@@ -4161,16 +4161,59 @@ def _decode(self, cur):
...
+        self.flow_info = {}
+        self.scope_id = {}
+        self.reserved_2 = {}
+        self.reserved_3 = {}
 
     def _decode(self, cur):
-        pass
-
+        i = 0
+        while True :
```
Comment (masenf) [inline src/pike/smb2.py:4197]: This class is the correct place to loop through all of the structures in the buffer... but these structures need to be stored as **separate objects**, not indexed into several dictionaries with the same key. No other structure in pike uses this style.

**Review by masenf — CHANGES_REQUESTED** [review]
This should be following MS-SMB2 structure descriptions.

You need a [SockaddrStorage structure](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-smb2/4b77102f-769f-414d-b137-47cabfe8be8f) that contains the IP fields.

You need a [NetworkInterfaceInfo](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-smb2/fcd862d1-1b85-42df-92b1-e103199f531f) frame that contains the other fields and references the sockaddr storage.

Finally, the QueryNetworkInterfaceInfoResponse should contain one or more NetworkInterfaceInfo objects as children.

If you're just wanting to hack it like this, I recommend implementing that in your test code, not in the upstream. This project aims to be as close to the protocol specification as possible.

File: src/pike/smb2.py [hunk 6]
```diff
@@ -4161,16 +4161,100 @@ def _decode(self, cur):
...
+     RSS_CAPABLE    = 0x00000001
+     RDMA_CAPABLE   = 0x00000002
+
+Capability.import_items(globals())
+
+class SockaddrIn:
+    def __init__(self):
+        port = array.array("B", [0] * 2)
+        ip = array.array("B", [0] * 4)
+        reserved = array.array("B", [0] * 8)
```
Comment (masenf) [inline src/pike/smb2.py:4181]:
```suggestion
        self.port = array.array("B", [0] * 2)
        self.ip = array.array("B", [0] * 4)
        self.reserved = array.array("B", [0] * 8) 
```

File: src/pike/smb2.py [hunk 7]
```diff
@@ -4161,16 +4161,100 @@ def _decode(self, cur):
...
+        self.port = cur.decode_uint16le()
+        self.ip = str(ipaddress.ip_address(cur.decode_uint32be())) 
+        self.reserved = cur.decode_uint64le()
+
+class SockaddrIn6:
+    def __init__(self):
+        port = array.array("B", [0] * 2)
+        flowinfo = array.array("B", [0] * 4)
+        ip = array.array("B", [0] * 16)
+        scopeid = array.array("B", [0] * 4)
```
Comment (masenf) [inline src/pike/smb2.py:4192]:
```suggestion
        self.port = array.array("B", [0] * 2)
        self.flowinfo = array.array("B", [0] * 4)
        self.ip = array.array("B", [0] * 16)
        self.scopeid = array.array("B", [0] * 4)
```

File: src/pike/smb2.py [hunk 8]
```diff
@@ -4161,16 +4161,100 @@ def _decode(self, cur):
...
+                self.buffer = sockaddr
+                self.reserved = array.array("B", [0] * (112))
+            elif isinstance(sockaddr, SockaddrIn6):
+                self.family= AddressFamily.IPv6 
+                self.buffer = sockaddr
+                self.reserved = array.array("B", [0] * (100))
+
+    def decode(self, cur):
+        self.family = AddressFamily(cur.decode_uint16le())
+        if self.family == IPv4 :
```
Comment (masenf) [inline src/pike/smb2.py:4217]:
```suggestion
        if self.family == IPv4:
```

File: src/pike/smb2.py [hunk 9]
```diff
@@ -4161,16 +4161,100 @@ def _decode(self, cur):
...
+            sockaddr = SockaddrIn()
+            sockaddr.decode(cur)
+            self.buffer = sockaddr
+            self.reserved = cur.decode_bytes(112)
+        elif self.family == IPv6 :
```
Comment (masenf) [inline src/pike/smb2.py:4222]:
```suggestion
        elif self.family == IPv6:
```

File: src/pike/smb2.py [hunk 10]
```diff
@@ -4161,16 +4161,100 @@ def _decode(self, cur):
...
+            sockaddr = SockaddrIn6()
+            sockaddr.decode(cur)
+            self.buffer = sockaddr
+            self.reserved = cur.decode_bytes(100)
+    
+class NetworkInterfaceInfo(core.Frame):
+    def __init__(self, parent):
+        core.Frame.__init__(self, parent)
```
Comment (masenf) [inline src/pike/smb2.py:4230]:
```suggestion
        super().__init__(parent)
```

File: src/pike/smb2.py [hunk 11]
```diff
@@ -4161,16 +4161,100 @@ def _decode(self, cur):
...
     ioctl_ctl_code = FSCTL_QUERY_NETWORK_INTERFACE_INFO
 
     def __init__(self, parent):
         IoctlOutput.__init__(self, parent)
+        self.ninfo = []
 
     def _decode(self, cur):
-        pass
-
+        while True :
```
Comment (masenf) [inline src/pike/smb2.py:4251]:
```suggestion
        while True:
```

File: src/pike/smb2.py [hunk 12]
```diff
@@ -4161,16 +4161,100 @@ def _decode(self, cur):
...
+        self.family = array.array("B", [0] * 2)
+        self.buffer = []
+        self.reserved = []
+        if sockaddr is not None:
+            if isinstance(sockaddr, SockaddrIn):
+                self.family = AddressFamily.IPv4
+                self.buffer = sockaddr
+                self.reserved = array.array("B", [0] * (112))
+            elif isinstance(sockaddr, SockaddrIn6):
+                self.family= AddressFamily.IPv6
```
Comment (masenf) [inline src/pike/smb2.py:4211]:
```suggestion
                self.family = AddressFamily.IPv6 
```

File: src/pike/smb2.py [hunk 13]
```diff
@@ -4161,16 +4161,100 @@ def _decode(self, cur):
...
+        scopeid = array.array("B", [0] * 4)
+    def decode(self, cur):
+        self.port = cur.decode_uint16le()
+        self.flowinfo = cur.decode_uint32le()
+        self.ip = binascii.hexlify(cur.decode_bytes(16))
+        self.scopeid = cur.decode_uint32le()
+
+class SockaddrStorage(core.Frame):
+    def __init__(self, parent, sockaddr=None):
+        core.Frame.__init__(self, parent)
```
Comment (masenf) [inline src/pike/smb2.py:4201]:
```suggestion
        super().__init__(parent)
```

File: src/pike/smb2.py [hunk 14]
```diff
@@ -4161,16 +4161,100 @@ def _decode(self, cur):
...
+        self.family = array.array("B", [0] * 2)
+        self.buffer = []
+        self.reserved = []
```
Comment (masenf) [inline src/pike/smb2.py:4204]: the types for these initial values seem a bit weird.

i would expect them to be initialized to an empty SockaddrIn and empty byte array.

**Review by masenf — CHANGES_REQUESTED** [review]
a few formatting changes are requested.

thanks for restructuring the decode in the pike style! that part looks great.

File: src/pike/smb2.py [hunk 15]
```diff
@@ -4161,16 +4161,100 @@ def _decode(self, cur):
...
+        self.scopeid = array.array("B", [0] * 4)
+    def decode(self, cur):
+        self.port = cur.decode_uint16le()
+        self.flowinfo = cur.decode_uint32le()
+        self.ip = binascii.hexlify(cur.decode_bytes(16))
+        self.scopeid = cur.decode_uint32le()
+
+class SockaddrStorage(core.Frame):
+    def __init__(self, parent, sockaddr=None):
+        super().__init__(self, parent)
```
Comment (masenf) [inline src/pike/smb2.py:4201]:
```suggestion
        core.Frame.__init__(self, parent)
```

File: src/pike/smb2.py [hunk 16]
```diff
@@ -4161,16 +4161,100 @@ def _decode(self, cur):
...
+            self.reserved = cur.decode_bytes(112)
+        elif self.family == IPv6:
+            sockaddr = SockaddrIn6()
+            sockaddr.decode(cur)
+            self.buffer = sockaddr
+            self.reserved = cur.decode_bytes(100)
+    
+class NetworkInterfaceInfo(core.Frame):
+    def __init__(self, parent):
+        super().__init__(self, parent)
```
Comment (masenf) [inline src/pike/smb2.py:4230]:
```suggestion
        core.Frame.__init__(self, parent)
```

**Review by masenf — CHANGES_REQUESTED** [review]
i forgot that we still support 2.7, which doesn't have `super()`, my bad.

**Review by masenf — COMMENTED** [review]
more py27 isms

**Comment by masenf** [comment]
The actual import error on py27 is

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/workspace/pike/.tox/py27/lib64/python2.7/site-packages/pike/__init__.py", line 31, in <module>
    from pike.test import TreeConnect as TreeConnect
  File "/workspace/pike/.tox/py27/lib64/python2.7/site-packages/pike/test/__init__.py", line 33, in <module>
    import pike.model as model
  File "/workspace/pike/.tox/py27/lib64/python2.7/site-packages/pike/model.py", line 51, in <module>
    from . import crypto
  File "/workspace/pike/.tox/py27/lib64/python2.7/site-packages/pike/crypto.py", line 32, in <module>
    from . import smb2
  File "/workspace/pike/.tox/py27/lib64/python2.7/site-packages/pike/smb2.py", line 35, in <module>
    import ipaddress
ImportError: No module named ipaddress
```

ipaddress isn't a module on py27 😔

will need to install the backport if this functionality is really needed.

**Review by isi-adas — APPROVED** [review]
LGTM, subject to passing CI.
Likely dependent on https://github.com/emc-isilon/pike/pull/132

File: src/pike/smb2.py [hunk 17]
```diff
@@ -4161,16 +4160,100 @@ def _decode(self, cur):
...
+class SockaddrIn6(object):
+    def __init__(self):
+        self.port = array.array("B", [0] * 2)
+        self.flowinfo = array.array("B", [0] * 4)
+        self.ip = array.array("B", [0] * 16)
+        self.scopeid = array.array("B", [0] * 4)
+    def decode(self, cur):
+        self.port = cur.decode_uint16le()
+        self.flowinfo = cur.decode_uint32le()
+        self.ip = ':'.join(str(ip_hex_raw[i:i+4]) for i in range(0, len(ip_hex_raw), 4))
```
[the remark this answers is not in the document]
Comment (sudosantanu) [reply src/pike/smb2.py:4196]: yes... we were not testing it initially, if tested we would have gotten error. This one can be done using  binascii.
Something like:
ip_hex = binascii.hexlify(cur).decode('utf-8')
self.ip  = ':'.join(ip_hex[i:i+4] for i in range(0, len(ip_hex), 4))

to keep the parity with IPV4, I did the IPv6 this way.

**Comment by masenf** [comment]
Rebased on master and cleaned up whitespace

**Review by isi-adas — APPROVED** [review]

**Pull request merged**